File: Server/CSharp/RagiSharkServerCS/WebSocketHeader.cs
namespace RagiSharkServerCS
{
    struct WebSocketHeader
    {
        public bool Fin;
        public byte Rcv1;
        public byte Rcv2;
        public byte Rcv3;
        public OpCode OpCode;
        public bool Mask;
        public int PayloadLength;

        public static WebSocketHeader Parse(byte[] bytes)
        {
            var header = new WebSocketHeader();
            if (bytes == null || bytes.Length < 2)
            {
                return header;
            }

            header.Fin = (bytes[0] >> 7) == 1;
            header.Rcv1 = (byte)(bytes[0] >> 6 & 1);
            header.Rcv2 = (byte)(bytes[0] >> 5 & 1);
            header.Rcv3 = (byte)(bytes[0] >> 4 & 1);
            header.OpCode = (OpCode)(bytes[0] & 0x0f);
            header.Mask = (bytes[1] >> 7) == 1;
            header.PayloadLength = bytes[1] & 0x7f;
            return header;
        }

        public static WebSocketHeader Create(bool fin, OpCode opCode, int payloadLength)
        {
            var header = new WebSocketHeader();
            header.Fin = fin;
            header.Rcv1 = 0;
            header.Rcv2 = 0;
            header.Rcv3 = 0;
            header.OpCode = opCode;
            header.Mask = false;
            header.PayloadLength = payloadLength;
            return header;
        }

        public byte[] ToBinary()
        {
            if (PayloadLength <= 125)
            {
                var bytes = new byte[2];
                bytes[0] = (byte)(((Fin ? 1 : 0) << 7) | (byte)OpCode);
                bytes[1] = (byte)PayloadLength;
                return bytes;
            }
            else if (PayloadLength <= 0xffff)
            {
                short len = (short)PayloadLength;
                var bytes = new byte[2 + 2];
                bytes[0] = (byte)(((Fin ? 1 : 0) << 7) | (byte)OpCode);
                bytes[1] = 126;
                bytes[2] = (byte)((len & 0xff00) >> 8);
                bytes[3] = (byte)(len & 0x00ff);
                return bytes;
            }
            else
            {
                long len = PayloadLength;
                var bytes = new byte[2 + 8];
                bytes[0] = (byte)(((Fin ? 1 : 0) << 7) | (byte)OpCode);
                bytes[1] = 127;
                bytes[2] = (byte)(len >> 56);
                bytes[3] = (byte)((len >> 48) & 0xff);
                bytes[4] = (byte)((len >> 40) & 0xff);
                bytes[5] = (byte)((len >> 32) & 0xff);
                bytes[6] = (byte)((len >> 24) & 0xff);
                bytes[7] = (byte)((len >> 16) & 0xff);
                bytes[8] = (byte)((len >> 8) & 0xff);
                bytes[9] = (byte)((len >> 0) & 0xff);
                return bytes;
            }
        }

        public override string ToString()
        {
            return $"fin: {Fin}, rcv1: {Rcv1}, rcv2: {Rcv2}, rcv3: {Rcv3}, opcode: {OpCode}, mask: {Mask}, payloadLength: {PayloadLength}";
        }
    }
}

File: Server/Swift/RagiSharkServerSwift/RagiSharkServerSwift/Sources/websocket/WebSocketHeader.swift
//
//  WebSocketHeader.swift
//  RagiSharkServerSwift
//
//  Created by ragingo on 2021/12/04.
//

import Foundation

struct WebSocketHeader {
    var fin: Bool = false
    var rcv1: UInt8 = 0
    var rcv2: UInt8 = 0
    var rcv3: UInt8 = 0
    var opCode: OpCode = .continuation
    var mask: Bool = false
    var payloadLength: Int = 0

    init() {}

    init(fin: Bool, opCode: OpCode, payloadLength: Int) {
        self.init()
        self.fin = fin
        self.opCode = opCode
        self.payloadLength = payloadLength
    }

    func toBinary() -> Data {
        let opCode = UInt8(opCode.rawValue)

        if payloadLength <= 125 {
            var bytes = Data(repeating: 0, count: 2)
            bytes[0] = UInt8(((fin ? 1 : 0) << 7) | opCode)
            bytes[1] = UInt8(payloadLength)
            return bytes
        }

        if payloadLength <= 0xffff {
            var bytes = Data(repeating: 0, count: 2 + 2)
            bytes[0] = UInt8(((fin ? 1 : 0) << 7) | opCode)
            bytes[1] = 126
            bytes[2] = UInt8((payloadLength & 0xff00) >> 8)
            bytes[3] = UInt8(payloadLength & 0x00ff)
            return bytes
        }

        var bytes = Data(repeating: 0, count: 2 + 8)
        bytes[0] = UInt8(((fin ? 1 : 0) << 7) | opCode)
        bytes[1] = 127
        bytes[2] = UInt8(payloadLength >> 56)
        bytes[3] = UInt8(payloadLength >> 48) & 0xff
        bytes[4] = UInt8(payloadLength >> 40) & 0xff
        bytes[5] = UInt8(payloadLength >> 32) & 0xff
        bytes[6] = UInt8(payloadLength >> 24) & 0xff
        bytes[7] = UInt8(payloadLength >> 16) & 0xff
        bytes[8] = UInt8(payloadLength >> 8) & 0xff
        bytes[9] = UInt8(payloadLength >> 0) & 0xff
        return bytes
    }

    static func parse(from data: Data) -> WebSocketHeader {
        var header = WebSocketHeader()
        if data.count < 2 {
            return header
        }
        let bytes = data.prefix(2)
        header.fin = (bytes[0] >> 7) == 1
        header.rcv1 = bytes[0] >> 6 & 1
        header.rcv2 = bytes[0] >> 5 & 1
        header.rcv3 = bytes[0] >> 4 & 1
        header.opCode = .init(rawValue: Int(bytes[0] & 0x0f)) ?? .continuation
        header.mask = (bytes[1] >> 7) == 1
        header.payloadLength = Int(bytes[1] & 0x7f)
        return header
    }
}

